dll/shellext/fontext/CFontExt.cpp

```diff
@@ -363,3 +365,161 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
     return S_OK;
 }
 
+// *** IDropTarget methods ***
+STDMETHODIMP CFontExt::DragEnter(IDataObject* pDataObj, DWORD grfKeyState, POINTL pt, DWORD* pdwEffect)
+{
+    *pdwEffect = DROPEFFECT_NONE;
+
+    CComHeapPtr<CIDA> cida;
+    HRESULT hr = _GetCidlFromDataObject(pDataObj, &cida);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+#if 1   // Please implement DoGetFontTitle
+    return DRAGDROP_S_CANCEL;
+#else
+    *pdwEffect = DROPEFFECT_COPY;
+    return S_OK;
+#endif
+}
+
+STDMETHODIMP CFontExt::DragOver(DWORD grfKeyState, POINTL pt, DWORD* pdwEffect)
+{
+    return S_OK;
+}
+
+STDMETHODIMP CFontExt::DragLeave()
+{
+    return S_OK;
+}
+
+STDMETHODIMP CFontExt::Drop(IDataObject* pDataObj, DWORD grfKeyState, POINTL pt, DWORD* pdwEffect)
+{
+    *pdwEffect = DROPEFFECT_NONE;
+
+    WCHAR szFontsDir[MAX_PATH];
+    HRESULT hr = SHGetFolderPathW(NULL, CSIDL_FONTS, NULL, 0, szFontsDir);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return E_FAIL;
+
+    CComHeapPtr<CIDA> cida;
+    hr = _GetCidlFromDataObject(pDataObj, &cida);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    WCHAR szRoot[MAX_PATH];
+    PCUIDLIST_ABSOLUTE pidlRoot = HIDA_GetPIDLFolder(cida);
+    SHGetPathFromIDListW(pidlRoot, szRoot);
+
+    BOOL bOK = TRUE;
+    CAtlArray<CStringW> FontPaths;
+    for (UINT n = 0; n < cida->cidl; ++n)
+    {
+        const FontPidlEntry* fontEntry = _FontFromIL(HIDA_GetPIDLItem(cida, n));
+        if (!fontEntry)
+            continue;
+
+        CStringW File = g_FontCache->Filename(fontEntry);
+        if (File.IsEmpty())
+        {
+            ERR("No file found for %S\n", fontEntry->Name);
+            continue;
+        }
+
+        if (PathIsRelativeW(File))
+        {
+            File = szRoot + File;
+        }
+
+        if (PathIsDirectoryW(File))
+        {
+            ERR("PathIsDirectory\n");
+            bOK = FALSE;
+            break;
+        }
+
+        LPCWSTR pchDotExt = PathFindExtensionW(File);
+        if (!IsFontDotExt(pchDotExt))
+        {
+            ERR("%S is not supported\n", pchDotExt);
+            bOK = FALSE;
+            break;
+        }
+
+        FontPaths.Add(File);
+    }
+
+    if (!bOK)
+        return E_FAIL;
+
+    HKEY hkeyFonts = NULL;
```

Comment:
If you use CRegKey, you can simply return failure codes in the loop below

Reply:
OK.